### ISSUE TITLE
building libomp for arm64

### ISSUE THREAD

**Issue by AllinCottrell** [tpoechtrager/osxcross#267]
I'm using current osxcross with MacOSX11.1.sdk, clang 11.1.0, and aarch64-apple-darwin20.2-clang seems to be working fine for most things, but my build of openmp-11.1.0 is failing. I got around an asm issue with a patch from macports, but at the linking stage I'm getting:

`Undefined symbols for architecture arm64:`
`  "___divdc3", referenced from:`
`      ___kmpc_atomic_cmplx8_div in kmp_atomic.cpp.o`
`      ___kmpc_atomic_cmplx10_div in kmp_atomic.cpp.o`
`      ___kmpc_atomic_cmplx4_div_cmplx8 in kmp_atomic.cpp.o`
`  "___divsc3", referenced from:`
`      ___kmpc_atomic_cmplx4_div in kmp_atomic.cpp.o`

My Toolchain file looks like:

`SET(CMAKE_SYSTEM_NAME Darwin)`
`SET(CMAKE_SYSTEM_VERSION 20.2)`
`SET(CMAKE_C_COMPILER   aarch64-apple-darwin20.2-clang)`
`SET(CMAKE_CXX_COMPILER aarch64-apple-darwin20.2-clang++)`
`SET(CMAKE_FIND_ROOT_PATH  /opt/mac11)`
`SET(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM BOTH)`
`SET(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)`
`SET(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE BOTH)`
`SET(CMAKE_OSX_SYSROOT /opt/mac11/SDK/MacOSX11.1.sdk)`

Any help would be much appreciated.

**Comment by Gcenx** [comment]
Checking the SDK `___divdc3` Is listed as x86_64-macos target only.

Seems something is screwy with your compile, to me it seems you’ve compiled for x86_64 but attempting to link as arm64.

**Comment by AllinCottrell** [comment]
Well, I'm just relying on cmake to get things right, via the Toolchain file noted above. I'm no expert in that department so maybe it's wrong/insufficient, or I'm referencing it wrongly. My setup command, in `openmp-11.1.0.src`, is

`cmake -DCMAKE_TOOLCHAIN_FILE=../Toolchain-arm64.cmake .`

where `Toolchain-arm64.cmake`has the content shown above, and lives one directory-level up, and I then follow with `make`.

**Comment by Gcenx** [comment]
I haven’t setup OSXCROSS for arm64 target currently since I have an actual M1 system.

From my testing in #265 its possible some more changes are required, alternatively libomp might be setting the needed flags incorrectly within the project.

Unless you need to compile libomp directly from source for a specific reason why not just use `osxcross-macports` and install the needed package?, I checked and I do see a prebuilt package for arm64 target.

**Comment by AllinCottrell** [comment]
Thanks, using osxcross-macports sounds like a good idea.

**Comment by Gcenx** [comment]
Just an fwi that still won’t guarantee your intended project will compile correctly for arm64. Both homebrew & Macports have been needing to apply additional fixes to get somethings to compile for the intended arch.

**Comment by AllinCottrell** [comment]
Yes, I'm kinda expecting some problems when my app is actually tested on M1.

BTW the current osxcross-macports script stops short at Darwin_19, but it's easy enough to add Darwin_20 support and I've now successfully installed the Macports libomp 11arm64 package in my osxcross tree.

**Comment by Gcenx** [comment]
My current issue with osxcross is xar and some other deps expect them to be installed into a fixed location over checking for there existence with pkg-config (OpenSSL being one)

Here’s a useful environment variable
```
# OSXCROSS will handle headers/libraries/frameworks injection
ENV OSXCROSS_MP_INC="YES"
```
That saved me a lot of headaches when previously working on phoenicis-winebuild.

For the moment I’ll lightly try messing with any issue sub-projects and provide patches to workaround any issues I find, since homebrew (most popular package manage for some reason) installs into /opt/homebrew on Apple silicon. Macports has always installed into /opt/local by default so no change for that.

**Comment by carlocab** [comment]
I seem to have come across this while working on https://github.com/Homebrew/homebrew-core/pull/79454. This is while compiling natively on arm64 while targeting arm64, so I have no idea how this is even coming up...

**Comment by Gcenx** [comment]
@carlocab I’d expectenced a similar issue when having brew build an x86_64 formula on an Apple Silicon Mac, avoiding the clang(++)/ld shims gotten the formula to build correctly. Also of note is Macports shows libomp 11.1.0 and 12.0.0 building for arm64 without issue.

**Comment by carlocab** [comment]
I'm trying to build with a newly-built compiler, so the shims are being bypassed here. The build with the shims actually works fine...

**Comment by Gcenx** [comment]
If you’ve kept your temp file better check the stage 1 omp was really built for aarch64 as it should like stage 1 was built for the wrong arch.

**Comment by carlocab** [comment]
Didn't build libomp in stage1, as there's no need for it when bootstrapping. 

I just meant that skipping a bootstrap compiler and building libomp with the shims works, but it seems now using a bootstrap compiler (and hence skipping the `clang[++]` shim) leads to a libomp failure.
